common/src/main/java/com/wynntils/wc/utils/ChatItemUtils.java

```diff
@@ -0,0 +1,247 @@
+import com.wynntils.core.webapi.profiles.item.ItemProfile;
+import com.wynntils.mc.mixin.accessors.ItemStackInfoAccessor;
+import com.wynntils.wc.custom.item.GearItemStack;
+import com.wynntils.wc.objects.ItemIdentificationContainer;
+import com.wynntils.wc.objects.Powder;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.TextComponent;
+import org.apache.commons.lang3.ArrayUtils;
+
+public class ChatItemUtils {
+
+    // private-use unicode chars
+    private static final String START = new String(Character.toChars(0xF5FF0));
+    private static final String END = new String(Character.toChars(0xF5FF1));
+    private static final String SEPARATOR = new String(Character.toChars(0xF5FF2));
+    private static final String RANGE =
+            "[" + new String(Character.toChars(0xF5000)) + "-" + new String(Character.toChars(0xF5F00)) + "]";
+    private static final int OFFSET = 0xF5000;
+
+    private static final boolean ENCODE_NAME = false;
+
+    public static final Pattern ENCODED_PATTERN = Pattern.compile(START + "(?<Name>.+?)" + SEPARATOR + "(?<Ids>" + RANGE
+            + "*)(?:" + SEPARATOR + "(?<Powders>" + RANGE + "+))?(?<Rerolls>" + RANGE + ")" + END);
+
+    /**
+     * Encodes the given item, as long as it is a standard gear item, into the following format
+     *
+     * START character (U+F5FF0)
+     * Item name (optionally encoded)
+     * SEPARATOR character (U+F5FF2)
+     * Identifications/stars (encoded)
+     * SEPARATOR (only if powdered)
+     * Powders (encoded) (only if powdered)
+     * Rerolls (encoded)
+     * END character (U+F5FF1)
+     *
+     * Any encoded "value" is added to the OFFSET character value U+F5000 and then converted into the corresponding Unicode character:
+     *
+     * The name is encoded based on the ASCII value of each character minus 32
+     *
+     * Identifications are encoded either as the raw value minus the minimum value of that ID, or if the range is larger than 100,
+     * the percent value 0 to 100 of the given roll.
+     * Regardless of either case, this number is multiplied by 4, and the number of stars present on that ID is added.
+     * This ensures that the value and star count can be encoded into a single character and be decoded later.
+     *
+     * Powders are encoded as numerical values 1-5. Up to 4 powders are encoded into a single character - for each new powder,
+     * the running total is multiplied by 6 before the new powder value is added. Thus, each individual powder can be decoded.
+     *
+     * Rerolls are simply encoded as a raw number.
+     *
```

Comment:
Maybe add something that this format is defined by Wynntils Legacy and can't be changed for backwards compatibility?

Reply:
I mean, we could change it in legacy and artemis both if we wanted to.